mutable_tree.go

```diff
@@ -1227,48 +1148,100 @@ func (tree *MutableTree) balance(node *Node, orphans *[]*Node) (newSelf *Node, e
 		}
 		if rightBalance <= 0 {
 			// Right Right Case
-			newNode, orphaned, err := tree.rotateLeft(node)
+			newNode, err := tree.rotateLeft(node)
 			if err != nil {
 				return nil, err
 			}
-			*orphans = append(*orphans, orphaned)
 			return newNode, nil
 		}
 		// Right Left Case
-		var rightOrphaned *Node
-
-		right, err := node.getRightNode(tree.ImmutableTree)
+		node.rightNodeKey = nil
+		node.rightNode, err = tree.rotateRight(rightNode)
 		if err != nil {
 			return nil, err
 		}
-		node.rightHash = nil
-		node.rightNodeKey = 0
-		node.rightNode, rightOrphaned, err = tree.rotateRight(right)
+		newNode, err := tree.rotateLeft(node)
 		if err != nil {
 			return nil, err
 		}
-		newNode, leftOrphaned, err := tree.rotateLeft(node)
-		if err != nil {
-			return nil, err
-		}
-
-		*orphans = append(*orphans, right, leftOrphaned, rightOrphaned)
 		return newNode, nil
 	}
 	// Nothing changed
 	return node, nil
 }
 
-func (tree *MutableTree) addOrphans(orphans []*Node) error {
-	for _, node := range orphans {
-		if !node.persisted {
-			// We don't need to orphan nodes that were never persisted.
-			continue
+// getOrphans gets orphaned nodes by the changes of the working tree.
+func (tree *MutableTree) getOrphans() []*NodeKey {
```

Comment:
I don't see how you can track the orphans try traverse the prev root?

Reply:
I guess you also need to clear those reference to avoid memory leak?
And the nodes come from `NodeDB`'s cache, could it be evicted by cache and load again from db? It feels not cleaner than previous solution to me.
If we can somehow track orphans by traversing new root and old root simultaneously, without "hacking* the old nodes, that'd be ideal.

Reply:
it could be, I will check it

Reply:
OK, it works, I updated it.